src/ledger/LedgerEntryScope.h

```diff
@@ -0,0 +1,453 @@
+    MACRO(ThreadParApply, GlobalParApply) \
+    MACRO(ThreadParApply, TxParApply) \
+    MACRO(TxParApply, ThreadParApply)
+
+namespace stellar
+{
+
+enum class StaticLedgerEntryScope : uint16_t
+{
+#define STATIC_SCOPE_MACRO(SCOPE_NAME) SCOPE_NAME,
+    FOREACH_STATIC_LEDGER_ENTRY_SCOPE(STATIC_SCOPE_MACRO)
+#undef STATIC_SCOPE_MACRO
+};
+
+template <StaticLedgerEntryScope> class LedgerEntryScopeID;
+
+}
+
+template <stellar::StaticLedgerEntryScope T>
+std::ostream& operator<<(std::ostream& os,
+                         stellar::LedgerEntryScopeID<T> const& obj);
+
+namespace stellar
+{
+
+// LedgerEntryScopeID uniquely identifies a scope for ledger entries. It is
+// space-optimized to a single 64 bit word since these are stuck on all
+// ScopedLedgerEntry instances and there might be a lot of those.
+template <StaticLedgerEntryScope S> class LedgerEntryScopeID
+{
+  public:
+    static constexpr StaticLedgerEntryScope staticScope = S;
+    uint16_t const mIndex;
+    uint32_t const mLedger;
+
+    // NB: index is limited to 16 bits to keep size down;
+    // the constructor will throw if you exceed that, but
+    // in practice scopes should always be tiny numbers like
+    // the thread cluster index or the bucketlist level.
+    LedgerEntryScopeID(size_t index, uint32_t ledger);
+
+    bool operator==(LedgerEntryScopeID const& other) const;
+    bool operator!=(LedgerEntryScopeID const& other) const;
+
+    friend std::ostream& ::operator<<(std::ostream& os,
+                                      LedgerEntryScopeID const& obj);
+};
+
+static_assert(
+    sizeof(LedgerEntryScopeID<StaticLedgerEntryScope::GlobalParApply>) == 8,
+    "Unexpected size for LedgerEntryScopeID");
+
+template <StaticLedgerEntryScope T> class LedgerEntryScope;
+template <StaticLedgerEntryScope S> class ScopedLedgerEntryOpt;
+
+template <StaticLedgerEntryScope S> class ScopedLedgerEntry
+{
+    static constexpr StaticLedgerEntryScope staticScope = S;
+    using ScopeIdT = LedgerEntryScopeID<S>;
+    using ScopeT = LedgerEntryScope<S>;
+    LedgerEntry mEntry;
+
+    friend class LedgerEntryScope<S>;
+    friend class ScopedLedgerEntryOpt<S>;
+#define FRIEND_MACRO(SCOPE) \
+    friend class LedgerEntryScope<StaticLedgerEntryScope::SCOPE>;
+    FOREACH_STATIC_LEDGER_ENTRY_SCOPE(FRIEND_MACRO)
+#undef FRIEND_MACRO
+
+    ScopedLedgerEntry(ScopeIdT scopeID, LedgerEntry const& entry);
+    ScopedLedgerEntry(ScopeIdT scopeID, LedgerEntry&& entry);
+
+  public:
+    LedgerEntryScopeID<S> const mScopeID;
+
+    // Must construct with a scope.
+    ScopedLedgerEntry<S>() = delete;
+
+    ScopedLedgerEntry(ScopedLedgerEntry<S> const& other);
+    ScopedLedgerEntry(ScopedLedgerEntry<S>&& other);
+    ScopedLedgerEntry<S>& operator=(ScopedLedgerEntry<S> const& other);
+    ScopedLedgerEntry<S>& operator=(ScopedLedgerEntry<S>&& other);
+
+    LedgerEntry const& readInScope(LedgerEntryScope<S> const& scope) const;
+    LedgerEntry& modifyInScope(LedgerEntryScope<S> const& scope);
```

Comment:
Should we remove the `LedgerEntry&` modify variant, given we want to enforce the "safer" lambda form?